arbor-sim/arbor
# S expr

Move s expression parsing of strings, used for parsing region and locset expressions, from the Python wrapper to the main library.

CMake
* Add `ARBDEV_COLOR` option that forces gcc and clang to always output color output.

Move label parsing code from Python to arbor C++ library
* Replace `hopefully` type implemented in python headers with an `arb::util::expected`.
* Update `locset` and `region` constructors that take string arguments to
      parse strings as s-expressions or "quoted" labels.
* updated the input stream modifier used to parse asc files to use a lookup table
      for substitutions, renamed it transmogrifier.                                                             

Tests:
* moved s expression parsing tests into `unit` and extended them.

General label concepts
* Allow arbitrary strings in labels.
* Add `parse_region_expression` and `parse_locset_expression` functions
      alongside the existing `parse_label_expression` function for use when
      a region or locset is expected. These calls will promote a quoted string
      `"label"` to `(region "label")` or `(locset "label")` respectively.
* Add string literals for labels so that the C++ interface can use 
      `"soma"_lab` instead of awkward escaping `"\"soma\""`.

Python:
* Remove old label parsing code.
* Simplify Python wrapper code.

General:
* replaced some `arb::util::any` and `arb::util::variant` instances with their `std::` version.

Fixes #1139
Fixes #1132
